### ISSUE TITLE
Publish the ssz crate to crates.io via CI

### ISSUE THREAD

**Issue by paulhauner** [sigp/lighthouse#2259]
## Description

I own the [`eth2_ssz`](https://crates.io/crates/eth2_ssz) crate which has a very outdated version of SSZ.

It would be nice to add something to CI to ensure that we keep that crate updated whenever we merge to `stable`. I could quite easily create an additional owner for the crate and then add their credentials to CI.

**Comment by michaelsproul** [comment]
One difficulty with auto-publishing crates is determining when they've changed "enough" to require a version bump/republish. And I think the version bump needs to be done manually (it needs to already be set in the `Cargo.toml` before attempting to publish). So far the best I can come up with is this:

1. Have a job on CI that detects code changes to the crate in question and _fails_ if the code is changed without the version being changed. The choice of version change is up to the committer, could be patch/minor/major. We could have scripts to assist with this version bumping process.
2. When code is merged to `stable`, publish any crate whose published version is older than its current version.

One flaw in this scheme is that it might force version skips (unless we publish upon any merge to `unstable`). e.g. change code, bump to `v1.1.1`, change again, bump to `v1.1.2`, merge to `stable`: published version jumps `v1.1.0` to `v1.1.2`.

**Comment by paulhauner** [comment]
Hey @realbigsean, would you be interested in this one when you have some spare time? We'd probably want to publish `ssz_types` as well.

**Comment by paulhauner** [comment]
Also, I wonder if @sorpaas would be open to us taking over the [`ssz`](https://crates.io/crates/ssz) crate? :pray: I get the feeling that it's not being updated anymore and I'm not sure if it's compatible with the production version of the SSZ spec.

**Comment by realbigsean** [comment]
Can do!

**Comment by michaelsproul** [comment]
If anyone is interested in working around this in the meantime, you can do so using `patch` and `git` dependencies:

```toml
[package]
name = "my_package"
version = "0.1.0"
edition = "2018"

[dependencies]
types = { git = "https://github.com/sigp/lighthouse", branch = "stable" }

[patch]
[patch.crates-io]
tree_hash = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
tree_hash_derive = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
eth2_ssz = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
eth2_ssz_derive = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
eth2_ssz_types = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
eth2_hashing = { git = "https://github.com/sigp/lighthouse", branch = "stable" }

```

**Comment by leruaa** [comment]
The workaround above works fine for `types`, but I'm having trouble doing the same thing for `eth2` crate: 
```
types = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
eth2 = { git = "https://github.com/sigp/lighthouse", branch = "stable" }

[patch]
[patch.crates-io]
tree_hash = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
tree_hash_derive = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
eth2_ssz = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
eth2_ssz_derive = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
eth2_ssz_types = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
eth2_hashing = { git = "https://github.com/sigp/lighthouse", branch = "stable" }
```

I have the following errors:
```
error[E0432]: unresolved import `k256::ecdsa`
 --> C:\...\.cargo\git\checkouts\discv5-7af694118de5b208\02d2c89\src\handler\crypto\ecdh.rs:3:5
  |
3 |     ecdsa::{SigningKey, VerifyingKey},
  |     ^^^^^ help: a similar path exists: `super::k256::ecdsa`

   Compiling hyper-tls v0.5.0
   Compiling types v0.2.0 (https://github.com/sigp/lighthouse?branch=stable#3b600acd)
error[E0277]: the trait bound `ThreadRng: enr::k256::elliptic_curve::rand_core::CryptoRng` is not satisfied
  --> C:\...\.cargo\git\checkouts\discv5-7af694118de5b208\02d2c89\src\handler\crypto\mod.rs:55:64
   |
55 |                 let ephem_sk = k256::ecdsa::SigningKey::random(rand::thread_rng());
   |                                                                ^^^^^^^^^^^^^^^^^^ the trait `enr::k256::elliptic_curve::rand_core::CryptoRng` is not implemented for `ThreadRng`
   |
  ::: C:\...\.cargo\registry\src\github.com-1ecc6299db9ec823\k256-0.8.1\src\ecdsa\sign.rs:38:29
   |
38 |     pub fn random(rng: impl CryptoRng + RngCore) -> Self {
   |                             --------- required by this bound in `SigningKey::random`

error[E0277]: the trait bound `ThreadRng: enr::k256::elliptic_curve::rand_core::RngCore` is not satisfied
  --> C:\...\.cargo\git\checkouts\discv5-7af694118de5b208\02d2c89\src\handler\crypto\mod.rs:55:64
   |
55 |                 let ephem_sk = k256::ecdsa::SigningKey::random(rand::thread_rng());
   |                                                                ^^^^^^^^^^^^^^^^^^ the trait `enr::k256::elliptic_curve::rand_core::RngCore` is not implemented for `ThreadRng`
   |
  ::: C:\...\.cargo\registry\src\github.com-1ecc6299db9ec823\k256-0.8.1\src\ecdsa\sign.rs:38:41
   |
38 |     pub fn random(rng: impl CryptoRng + RngCore) -> Self {
   |                                         ------- required by this bound in `SigningKey::random`
```
What am I missing?

**Comment by realbigsean** [comment]
I tried this out locally and it seems like `enr` dependency is being updated from `0.5.0` to `0.5.1`. And even though the versioning suggests this should be non-breaking, it does break things.  I think because its dependency `k256` was updated from `0.7` to `0.8` 

For me, adding this to the `[dependencies]` section resolved it: 

`enr = { version = "=0.5.0", features = ["ed25519", "k256"] }`

Not sure why we don't see this in lighthouse honestly

**Comment by leruaa** [comment]
Thanks!

I was able to solve it by cloning and patching the `discv5` latest version:
```
[patch.'https://github.com/sigp/discv5']
discv5 = { path = "../../discv5", features = ["libp2p"] } 
```

**Comment by realbigsean** [comment]
> Have a job on CI that detects code changes to the crate in question and fails if the code is changed without the version being changed. The choice of version change is up to the committer, could be patch/minor/major.

@michaelsproul I've made a first pass at implementing this in #2423

> When code is merged to stable, publish any crate whose published version is older than its current version.

I have not implemented this. As a first step though, I've made the CI check also verify that if a crate's version is updated in the PR, the updated version exists on crates.io. So the CI task will fail until a manual publish happens.  

> One flaw in this scheme is that it might force version skips (unless we publish upon any merge to unstable). e.g. change code, bump to v1.1.1, change again, bump to v1.1.2, merge to stable: published version jumps v1.1.0 to v1.1.2.

I'd be in favor of publishes on merges into `unstable`.  I think it makes sense to separate these crates from the lighthouse release cycle.

**Comment by paulhauner** [comment]
> I'd be in favor of publishes on merges into unstable. I think it makes sense to separate these crates from the lighthouse release cycle.

This seems reasonable to me, but it would be nice to be able to optionally merge to `unstable` *without* updating crates.io. I see two options and I prefer (2) since it's simple and flexible.

### 1. Don't publish if there's an `.rc` suffix

 CI won't push to crates.io if there's a `-rc.x` version suffix? With that we could have `unstable` history look like this:

- `ssz v1.0.0`: CI publishes to crates.io
- `ssz v1.1.0-rc.0`: CI *does not* publish to crates.io
- `ssz v1.1.0`:  CI publishes to crates.io

### 2. Don't publish if there's an ignore file present

Another option would be to use a `.crates-io-ignore` file when we don't want to push a version yet. E.g, CI will never publish the `eth2_ssz` crate if `consensus/eth2_ssz/.crates-io-ignore` exists.

**Comment by realbigsean** [comment]
I took a peek at the [tokio repo](https://github.com/tokio-rs/tokio), and looks like they tag full releases with this syntax: `tokio-1.2.0`, and also tag commits associated with individual crates: `tokio-streams-0.6.7`. So we could also automatically publish to crates.io when pushing a tag to `unstable` referencing one of the crates?  And we could make the version check a part of CI when merging into `stable` from release branches.  

The version check would make sure any code changes in a relevant crate being merged into `stable` would be accompanied by a check to:
  - Make sure the version of the crate is also updated
  - Make sure the updated version exists on crates.io

I could see it being an annoying last step before a release if we forgot to publish crates. But I think it'd be worth having.

**Comment by realbigsean** [comment]
I spoke with @paulhauner offline about this. He agreed about auto-publishing crates when we push `<crate-name>-<version>` tags to `unstable`.  He made a good point about the version check being unnecessary if we go this route. Because in order to use an updated crate in `lighthouse`, we'll have have to make a second lighthouse PR updating the crate's version and making any necessary changes. Because we won't be relying on `[patch]` anymore.

An open question we have still is whether to move all `ssz` related crates into another repo. Wondering if you have any thoughts @michaelsproul?

**Comment by michaelsproul** [comment]
> An open question we have still is whether to move all ssz related crates into another repo. Wondering if you have any thoughts @michaelsproul?

Auto-publishing based on tags sounds good to me :+1: The two step workflow of updating the underlying crate and then updating Lighthouse itself sounds fine. I think we can hold off on separate repos for now, but could move to that easily enough if there's friction.

**Comment by sorpaas** [comment]
> Also, I wonder if @sorpaas would be open to us taking over the ssz crate? 🙏 I get the feeling that it's not being updated anymore and I'm not sure if it's compatible with the production version of the SSZ spec.

Sorry I missed this message. Apparently late to the party as the name has decided to be `eth2_ssz`.

I'll be resurrecting the `ssz` crate though. Move it to `rust-ethereum` org and add a group owner so we get more people with push rights. If there are still ways to collaborate please let me know!

**Comment by paulhauner** [comment]
Resolved via #2472
